File: products/windshift/internal/routes/ai.go
package routes

import (
	"net/http"

	"windshift/internal/handlers"
)

// RegisterAIRoutes registers AI-powered feature routes.
func RegisterAIRoutes(deps *Deps) {
	api := deps.API
	auth := deps.AuthMiddleware.RequireAuth
	admin := deps.PermissionMiddleware.RequireSystemAdmin()

	api.HandleH("GET /agent-studio/openapi.yaml", auth(http.HandlerFunc(handlers.AgentStudioOpenAPIYAML)))

	// AI feature endpoints (user) - rate limited to protect expensive LLM calls
	api.HandleH("GET /ai/status", auth(http.HandlerFunc(deps.AI.AI.Status)))
	api.HandleH("POST /ai/chat", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.Chat))))
	api.HandleH("GET /ai/sessions", auth(http.HandlerFunc(deps.AI.AI.ListAgentSessions)))
	api.HandleH("GET /ai/sessions/general", auth(http.HandlerFunc(deps.AI.AI.GetGeneralSession)))
	api.HandleH("GET /ai/sessions/{id}/messages", auth(http.HandlerFunc(deps.AI.AI.ListAgentMessages)))
	api.HandleH("POST /ai/sessions/{id}/archive", auth(http.HandlerFunc(deps.AI.AI.ArchiveAgentSession)))
	api.HandleH("GET /workspaces/{workspaceId}/available-standard-agents", auth(http.HandlerFunc(deps.AI.AI.ListAvailableStandardAgents)))
	api.HandleH("POST /workspaces/{workspaceId}/agent-sessions", auth(http.HandlerFunc(deps.AI.AI.CreateStandardSession)))
	api.HandleH("GET /ai/daily-briefing", auth(http.HandlerFunc(deps.AI.AI.GetDailyBriefing)))
	api.HandleH("GET /ai/plan-my-day", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.PlanMyDay))))
	api.HandleH("POST /ai/items/{id}/catch-me-up", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.CatchMeUp))))
	api.HandleH("POST /ai/items/{id}/find-similar", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.FindSimilarItems))))
	api.HandleH("POST /ai/items/{id}/decompose", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.DecomposeItem))))
	api.HandleH("POST /ai/milestones/{id}/generate-release-notes", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.GenerateReleaseNotes))))
	api.HandleH("POST /ai/iterations/{id}/analyze-dependencies", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.AnalyzeDependencies))))
	api.HandleH("POST /ai/iterations/{id}/accept-dependencies", auth(http.HandlerFunc(deps.AI.AI.AcceptDependencies)))
	api.HandleH("POST /ai/test-sets/{id}/summarize-description", auth(deps.AIRateLimiter.Limit(http.HandlerFunc(deps.AI.AI.SummarizeTestPlanDescription))))

	// LLM provider info (user)
	api.HandleH("GET /llm/providers", auth(http.HandlerFunc(deps.AI.LLMConnection.GetProviders)))
	api.HandleH("GET /llm/connections", auth(http.HandlerFunc(deps.AI.LLMConnection.GetEnabledConnections)))

	// LLM connection management (admin)
	api.HandleH("GET /admin/llm-connections", admin(http.HandlerFunc(deps.AI.LLMConnection.ListConnections)))
	api.HandleH("POST /admin/llm-connections", admin(http.HandlerFunc(deps.AI.LLMConnection.CreateConnection)))
	api.HandleH("GET /admin/llm-connections/{id}", admin(http.HandlerFunc(deps.AI.LLMConnection.GetConnection)))
	api.HandleH("PUT /admin/llm-connections/{id}", admin(http.HandlerFunc(deps.AI.LLMConnection.UpdateConnection)))
	api.HandleH("DELETE /admin/llm-connections/{id}", admin(http.HandlerFunc(deps.AI.LLMConnection.DeleteConnection)))
	api.HandleH("POST /admin/llm-connections/{id}/test", admin(http.HandlerFunc(deps.AI.LLMConnection.TestConnection)))
	api.HandleH("POST /admin/llm/providers/{type}/refresh-models", admin(http.HandlerFunc(deps.AI.LLMConnection.RefreshProviderModels)))
	api.HandleH("GET /admin/work-item-staleness", admin(http.HandlerFunc(deps.AI.WorkItemStaleness.Get)))
	api.HandleH("PUT /admin/work-item-staleness", admin(http.HandlerFunc(deps.AI.WorkItemStaleness.Update)))
}

File: products/windshift/frontend/src/lib/api/ai.js
import { del, get, post, put } from './core.js';

export const ai = {
  status: () => get('/ai/status'),
  planMyDay: (connectionId) =>
    get(`/ai/plan-my-day${connectionId ? `?connection_id=${connectionId}` : ''}`),
  planMyDayPreview: () => get('/ai/plan-my-day?preview=true'),
  catchMeUp: (itemId) => post(`/ai/items/${itemId}/catch-me-up`),
  findSimilar: (itemId) => post(`/ai/items/${itemId}/find-similar`),
  decompose: (itemId) => post(`/ai/items/${itemId}/decompose`),
  generateReleaseNotes: (milestoneId, connectionId) =>
    post(
      `/ai/milestones/${milestoneId}/generate-release-notes${connectionId ? `?connection_id=${connectionId}` : ''}`
    ),
  summarizeTestPlanDescription: (testSetId, connectionId) =>
    post(
      `/ai/test-sets/${testSetId}/summarize-description${connectionId ? `?connection_id=${connectionId}` : ''}`
    ),
  analyzeDependencies: (iterationId, body = {}, connectionId) =>
    post(
      `/ai/iterations/${iterationId}/analyze-dependencies${connectionId ? `?connection_id=${connectionId}` : ''}`,
      body
    ),
  acceptDependencies: (iterationId, suggestions) =>
    post(`/ai/iterations/${iterationId}/accept-dependencies`, { suggestions }),
  chat: (message, connectionId, sessionId, context) =>
    post('/ai/chat', {
      message,
      ...(connectionId ? { connection_id: connectionId } : {}),
      ...(sessionId ? { session_id: sessionId } : {}),
      ...(context && Object.keys(context).length ? { context } : {}),
    }),
  getGeneralSession: () => get('/ai/sessions/general'),
  listSessions: (includeArchived = false) =>
    get(`/ai/sessions${includeArchived ? '?include_archived=true' : ''}`),
  getSessionMessages: (sessionId) => get(`/ai/sessions/${sessionId}/messages`),
  archiveSession: (sessionId) => post(`/ai/sessions/${sessionId}/archive`),
  listAvailableStandardAgents: (workspaceId) =>
    get(`/workspaces/${workspaceId}/available-standard-agents`),
  createStandardSession: (workspaceId, data) =>
    post(`/workspaces/${workspaceId}/agent-sessions`, data),
  dailyBriefing: () => get('/ai/daily-briefing'),
};

export const aiFeatures = {
  getConfig: () => get('/admin/ai-features'),
  updateConfig: (data) => put('/admin/ai-features', data),
};

export const workItemStaleness = {
  get: () => get('/admin/work-item-staleness'),
  update: (data) => put('/admin/work-item-staleness', data),
};

export const llmConnections = {
  getAll: () => get('/admin/llm-connections'),
  get: (id) => get(`/admin/llm-connections/${id}`),
  create: (data) => post('/admin/llm-connections', data),
  update: (id, data) => put(`/admin/llm-connections/${id}`, data),
  delete: (id) => del(`/admin/llm-connections/${id}`),
  test: (id) => post(`/admin/llm-connections/${id}/test`),
};

export const llmProviders = {
  getProviders: () => get('/llm/providers'),
  getEnabled: () => get('/llm/connections'),
  refreshModels: (type, options = {}) =>
    post(`/admin/llm/providers/${encodeURIComponent(type)}/refresh-models`, options),
};

export const actionCapabilities = {
  getAll: () => get('/admin/action-capabilities'),
  get: (id) => get(`/admin/action-capabilities/${id}`),
  create: (data) => post('/admin/action-capabilities', data),
  update: (id, data) => put(`/admin/action-capabilities/${id}`, data),
  delete: (id) => del(`/admin/action-capabilities/${id}`),
  // Workspace-scoped picker list — returns enabled capabilities the workspace
  // can reference (applies-to-all OR explicitly scoped). Optional type filter.
  getForWorkspace: (workspaceId, type) => {
    const qs = type ? `?type=${encodeURIComponent(type)}` : '';
    return get(`/workspaces/${workspaceId}/action-capabilities${qs}`);
  },
};

// Runner pools (WI-177). A runner_pool is an ActionCapability; these manage its
// child resources — registration tokens (mint/list/revoke) and registered
// runner instances (list/revoke). The plaintext token is returned once on mint.
export const runnerPools = {
  listTokens: (capabilityId) => get(`/admin/action-capabilities/${capabilityId}/runner-tokens`),
  mintToken: (capabilityId, data = {}) =>
    post(`/admin/action-capabilities/${capabilityId}/runner-tokens`, data),
  revokeToken: (capabilityId, tokenId) =>
    del(`/admin/action-capabilities/${capabilityId}/runner-tokens/${tokenId}`),
  listInstances: (capabilityId) =>
    get(`/admin/action-capabilities/${capabilityId}/runner-instances`),
  revokeInstance: (capabilityId, instanceId) =>
    del(`/admin/action-capabilities/${capabilityId}/runner-instances/${instanceId}`),
  listWorkspaceTokens: (workspaceId, poolId) =>
    get(`/workspaces/${workspaceId}/agent-runner-pools/${poolId}/tokens`),
  mintWorkspaceToken: (workspaceId, poolId, data = {}) =>
    post(`/workspaces/${workspaceId}/agent-runner-pools/${poolId}/tokens`, data),
  revokeWorkspaceToken: (workspaceId, poolId, tokenId) =>
    del(`/workspaces/${workspaceId}/agent-runner-pools/${poolId}/tokens/${tokenId}`),
  listWorkspaceInstances: (workspaceId, poolId) =>
    get(`/workspaces/${workspaceId}/agent-runner-pools/${poolId}/instances`),
};

// actionCredentials: workspace-aware credential store referenced by HTTP
// capabilities. The plaintext secret travels only on create/rotate; every
// response is the sanitized DTO (has_secret + prefix; no ciphertext).
export const actionCredentials = {
  // Admin view — system-admin only. The create/update body chooses scope via
  // applies_to_all_workspaces + workspace_ids; the list returns every row.
  getAllGlobal: () => get('/admin/action-credentials'),
  createGlobal: (data) => post('/admin/action-credentials', data),
  updateGlobal: (id, data) => put(`/admin/action-credentials/${id}`, data),
  rotateGlobal: (id, secret) => post(`/admin/action-credentials/${id}/rotate`, { secret }),
  deleteGlobal: (id) => del(`/admin/action-credentials/${id}`),

  // Workspace-scoped credentials — gated by action.credential.manage. The
  // list returns credentials usable in the workspace (those that apply to all
  // workspaces, plus those scoped to this one).
  getForWorkspace: (workspaceId) => get(`/workspaces/${workspaceId}/action-credentials`),
  createForWorkspace: (workspaceId, data) =>
    post(`/workspaces/${workspaceId}/action-credentials`, data),
  updateForWorkspace: (workspaceId, id, data) =>
    put(`/workspaces/${workspaceId}/action-credentials/${id}`, data),
  rotateForWorkspace: (workspaceId, id, secret) =>
    post(`/workspaces/${workspaceId}/action-credentials/${id}/rotate`, { secret }),
  deleteForWorkspace: (workspaceId, id) =>
    del(`/workspaces/${workspaceId}/action-credentials/${id}`),
};
